test-support/page-object/helpers.js

```diff
@@ -1,26 +1,120 @@
 import Ember from 'ember';
+import Ceibo from 'ceibo';
 
-export function qualifySelector(...selectors) {
-  return selectors.filter(item => !!item).join(' ');
+var { trim } = Ember.$;
+
+class Selector {
+  constructor(node, scope, selector, filters) {
+    this.targetNode = node;
+    this.targetScope = scope || '';
+    this.targetSelector = selector || '';
+    this.targetFilters = filters;
+  }
+
+  toString() {
+    var scope,
+        filters;
+
+    if (this.targetFilters.resetScope) {
+      scope = this.targetScope;
+    } else {
+      scope = this.calculateScope(this.targetNode, this.targetScope);
+    }
+
+    filters = this.calculateFilters(this.targetFilters);
+
+    return trim(`${scope} ${this.targetSelector}${filters}`);
+  }
+
+  calculateFilters() {
+    var filters = [];
+
+    if (this.targetFilters.contains) {
+      filters.push(`:contains("${this.targetFilters.contains}")`);
+    }
+
+    if (typeof this.targetFilters.at === 'number') {
+      filters.push(`:eq(${this.targetFilters.at})`);
+    } else if (this.targetFilters.last) {
+      filters.push(':last');
+    }
+
+    return filters.join('');
+  }
+
+  calculateScope(node, targetScope) {
+    var scopes = this.getScopes(node);
+
+    scopes.reverse();
+    scopes.push(targetScope);
+
+    return $.trim(scopes.join(' '));
+  }
+
+  getScopes(node) {
+    var scopes = [];
+
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: ['target' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963753/issues/source?bid=2903707&fileBranchId=3076256#l56)